routing/payment_lifecycle.go

```diff
@@ -66,158 +67,243 @@ type shardOutcome struct {
+
+	// If we had any existing attempts outstanding, we'll start by and
+	// resume spinning up goroutines that'll collect their results and
+	// deliver them to the lifecycle loop below.
+	existingAttempts, err := p.router.cfg.Control.GetAttempts(
+		p.paymentHash,
 	)
+	if err != nil {
+		return [32]byte{}, nil, err
+	}
 
-	// If we have an existing attempt, we'll start by collecting its result.
-	if p.existingAttempt != nil {
-		attempt = p.existingAttempt
+	for _, a := range existingAttempts {
+		attempt := a
 
-		var err error
-		outcome, err = shardHandler.collectResult(attempt)
-		if err != nil {
-			return [32]byte{}, nil, err
-		}
+		shardHandler.wg.Add(1)
+		go func() {
+			defer shardHandler.wg.Done()
+
+			outcome, err := shardHandler.collectResult(
+				&attempt.HTLCAttemptInfo,
+			)
+			if err != nil {
+				select {
+				case criticalErrs <- err:
+				case <-shardHandler.quit:
+				}
+				return
+			}
+
+			select {
+			case shardOutcomes <- outcome:
+			case <-shardHandler.quit:
+			}
+		}()
 	}
 
+	var (
+		success         = false
+		routeFailure    *paymentFailure
+		terminalFailure *paymentFailure
+	)
+
 	// We'll continue until either our payment succeeds, or we encounter a
 	// critical error during path finding.
+Loop:
 	for {
 
-		// If this payment had no existing payment attempt, we create
-		// and send one now.
-		if attempt == nil {
+		// Check if we got any terminal failure.
+		var failure *paymentFailure
+		switch {
+		case terminalFailure != nil:
+			failure = terminalFailure
+
+		// We only consider the route failure if we didn't have another
+		// terminal failure.
+		case routeFailure != nil:
+			failure = routeFailure
+		}
+
+		// Get all attempts made for this payment, and extract the ones
+		// still in flight.
+		attempts, err := p.router.cfg.Control.GetAttempts(
+			p.paymentHash,
+		)
+		if err != nil {
+			return [32]byte{}, nil, err
+		}
+
+		var activeShards int
+		var sentAmt lnwire.MilliSatoshi
+		for _, a := range attempts {
+			if a.Settle != nil || a.Failure != nil {
+				continue
+			}
+
+			activeShards++
+			sentAmt += a.Route.Amt()
+		}
+
+		// The payment is finished if we have no outstanding shards,
+		// and we recorded either a success or a failure.
+		if activeShards == 0 && (success || failure != nil) {
+			// Find the first successful shard and return the
+			// preimage and route.
+			for _, a := range attempts {
+				if a.Settle != nil {
+					return a.Settle.Preimage, &a.Route, nil
+				}
+
+			}
+
+			// Payment failed, mark it as as permanently failed
+			// with the control tower.
+			if failure != nil {
+				saveErr := p.router.cfg.Control.Fail(
+					p.paymentHash, failure.failureCode,
+				)
+				if saveErr != nil {
+					return [32]byte{}, nil, saveErr
+				}
+
+				// Terminal state, return.
+				return [32]byte{}, nil, failure.err
+			}
+
+			return [32]byte{}, nil, fmt.Errorf("no successful " +
+				"attempts nor failure found")
+		}
+
+		remValue := p.totalAmount - sentAmt
```

Comment:
imo it'd be wise to panic on underflow here

Reply:
See https://github.com/lightningnetwork/lnd/pull/3970#discussion_r395328726